rust/src/graph/mod.rs

```diff
@@ -0,0 +1,137 @@
+    #[getset(get_copy = "pub")]
+    is_invisible: bool,
+
+    #[getset(get_copy = "pub")]
+    is_squashed: bool,
+}
+
+impl Module {
+    pub fn name_as_string(&self) -> String {
+        let interner = MODULE_NAMES.read().unwrap();
+        interner.resolve(self.name).unwrap().to_owned()
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct Graph {
+    // Hierarchy
+    modules_by_name: BiMap<DefaultSymbol, ModuleToken>,
+    modules: SlotMap<ModuleToken, Module>,
+    module_parents: SecondaryMap<ModuleToken, Option<ModuleToken>>,
+    module_children: SecondaryMap<ModuleToken, FxHashSet<ModuleToken>>,
+    // Imports
+    imports: SecondaryMap<ModuleToken, FxHashSet<ModuleToken>>,
+    reverse_imports: SecondaryMap<ModuleToken, FxHashSet<ModuleToken>>,
+    import_details: FxHashMap<(ModuleToken, ModuleToken), FxHashSet<ImportDetails>>,
+}
+
+impl From<ModuleToken> for Vec<ModuleToken> {
+    fn from(value: ModuleToken) -> Self {
+        vec![value]
+    }
+}
+
+impl From<ModuleToken> for FxHashSet<ModuleToken> {
+    fn from(value: ModuleToken) -> Self {
+        FxHashSet::from_iter([value])
+    }
+}
+
+pub trait ExtendWithDescendants:
+    Sized + Clone + IntoIterator<Item = ModuleToken> + Extend<ModuleToken>
+{
+    /// Extend this collection of module tokens with all descendant items.
+    fn extend_with_descendants(&mut self, graph: &Graph) {
+        for item in self.clone().into_iter() {
+            let descendants = graph.get_module_descendants(item).map(|item| item.token());
+            self.extend(descendants);
+        }
+    }
+
+    /// Extend this collection of module tokens with all descendant items.
+    fn with_descendants(mut self, graph: &Graph) -> Self {
+        self.extend_with_descendants(graph);
+        self
+    }
+}
+
+impl<T: Sized + Clone + IntoIterator<Item = ModuleToken> + Extend<ModuleToken>>
+    ExtendWithDescendants for T
+{
+}
+
+pub trait ModuleIterator<'a>: Iterator<Item = &'a Module> + Sized {
+    fn tokens(self) -> impl Iterator<Item = ModuleToken> {
+        self.map(|m| m.token)
+    }
+
+    fn names(self) -> impl Iterator<Item = DefaultSymbol> {
+        self.map(|m| m.name)
+    }
+
+    fn names_as_strings(self) -> impl Iterator<Item = String> {
+        let interner = MODULE_NAMES.read().unwrap();
+        self.map(move |m| interner.resolve(m.name).unwrap().to_owned())
+    }
+
+    fn visible(self) -> impl ModuleIterator<'a> {
+        self.filter(|m| !m.is_invisible)
+    }
+}
+
+impl<'a, T: Iterator<Item = &'a Module>> ModuleIterator<'a> for T {}
+
+pub trait ModuleTokenIterator<'a>: Iterator<Item = &'a ModuleToken> + Sized {
+    fn into_module_iterator(self, graph: &'a Graph) -> impl ModuleIterator<'a> {
+        self.map(|m| graph.get_module(*m).unwrap())
+    }
+}
+
+impl<'a, T: Iterator<Item = &'a ModuleToken>> ModuleTokenIterator<'a> for T {}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, new, Getters, CopyGetters)]
+pub struct ImportDetails {
```

Comment:
This is maybe using slightly different terminology to https://grimp.readthedocs.io/en/stable/usage.html#ImportGraph.get_import_details. I'm not massively happy with the name I chose for that method originally. But I think it might be worth coming up with a non-plural name for the struct here as it's more ergonomic. `ImportMetadata`?

Reply:
Yes true your suggestion is better 👍 Let's change it.